File: mpi/examples/mpi_struct.f
C ****************************************************************************
C FILE: mpi_struct.f
C DESCRIPTION:
C   MPI tutorial example code: Structure Derived Datatype
C AUTHOR: Blaise Barney
C LAST REVISED: 04/02/05
C ****************************************************************************

      program struct
      include 'mpif.h'

      integer NELEM
      parameter(NELEM=25)
      integer numtasks, rank, source, dest, tag, i,  ierr
      integer stat(MPI_STATUS_SIZE)

      type Particle
      sequence
      real*4 x, y, z, velocity
      integer n, type
      end type Particle

      type (Particle) p(NELEM), particles(NELEM)
      integer particletype, oldtypes(0:1), blockcounts(0:1), 
     &        offsets(0:1), extent

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

C     Setup description of the 4 MPI_REAL fields x, y, z, velocity 
      offsets(0) = 0
      oldtypes(0) = MPI_REAL
      blockcounts(0) = 4

C     Setup description of the 2 MPI_INTEGER fields n, type 
C     Need to first figure offset by getting size of MPI_REAL
      call MPI_TYPE_EXTENT(MPI_REAL, extent, ierr)
      offsets(1) = 4 * extent
      oldtypes(1) = MPI_INTEGER
      blockcounts(1) = 2

C     Now define structured type and commit it 
      call MPI_TYPE_STRUCT(2, blockcounts, offsets, oldtypes, 
     &                     particletype, ierr)
      call MPI_TYPE_COMMIT(particletype, ierr)
  
C     Initialize the particle array and then send it to each task
      tag = 1
      if (rank .eq. 0) then
         do 10 i=0, NELEM-1
         particles(i) = Particle ( 1.0*i, -1.0*i, 1.0*i, 
     &                  0.25, i, mod(i,2) )
  10     continue

         do 20 i=0, numtasks-1
         call MPI_SEND(particles, NELEM, particletype, i, tag, 
     &                 MPI_COMM_WORLD, ierr)
  20     continue
      endif

      source = 0
      call MPI_RECV(p, NELEM, particletype, source, tag, 
     &              MPI_COMM_WORLD, stat, ierr)

      print *, 'rank= ',rank,' p(3)= ',p(3)
      call MPI_TYPE_FREE(particletype, ierr)
      call MPI_FINALIZE(ierr)
      end

File: mpi/examples/mpi_struct.c
/******************************************************************************
* FILE: mpi_struct.c
* DESCRIPTION:
*   MPI tutorial example code: Structure Derived Datatype
* AUTHOR: Blaise Barney
* LAST REVISED: 04/13/05
******************************************************************************/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#define NELEM 25

int main (int argc, char *argv[])
{
int numtasks, rank, source=0, dest, tag=1, i;

typedef struct {
  float x, y, z;
  float velocity;
  int  n, type;
  }          Particle;
Particle     p[NELEM], particles[NELEM];
MPI_Datatype particletype, oldtypes[2];
int          blockcounts[2];

/* MPI_Aint type used to be consistent with syntax of */
/* MPI_Type_extent routine */
MPI_Aint    offsets[2], extent;

MPI_Status stat;

MPI_Init(&argc,&argv);
MPI_Comm_rank(MPI_COMM_WORLD, &rank);
MPI_Comm_size(MPI_COMM_WORLD, &numtasks);

/* Setup description of the 4 MPI_FLOAT fields x, y, z, velocity */
offsets[0] = 0;
oldtypes[0] = MPI_FLOAT;
blockcounts[0] = 4;

/* Setup description of the 2 MPI_INT fields n, type */
/* Need to first figure offset by getting size of MPI_FLOAT */
MPI_Type_extent(MPI_FLOAT, &extent);
offsets[1] = 4 * extent;
oldtypes[1] = MPI_INT;
blockcounts[1] = 2;

/* Now define structured type and commit it */
MPI_Type_struct(2, blockcounts, offsets, oldtypes, &particletype);
MPI_Type_commit(&particletype);

/* Initialize the particle array and then send it to each task */
if (rank == 0) {
  for (i=0; i<NELEM; i++) {
     particles[i].x = i * 1.0;
     particles[i].y = i * -1.0;
     particles[i].z = i * 1.0;
     particles[i].velocity = 0.25;
     particles[i].n = i;
     particles[i].type = i % 2;
     }
  for (i=0; i<numtasks; i++)
     MPI_Send(particles, NELEM, particletype, i, tag, MPI_COMM_WORLD);
  }

MPI_Recv(p, NELEM, particletype, source, tag, MPI_COMM_WORLD, &stat);

/* Print a sample of what was received */
printf("rank= %d   %3.2f %3.2f %3.2f %3.2f %d %d\n", rank,p[3].x,
     p[3].y,p[3].z,p[3].velocity,p[3].n,p[3].type);

MPI_Type_free(&particletype);
MPI_Finalize();
}
